ambry-commons/src/main/java/com/github/ambry/commons/FutureUtils.java

```diff
@@ -92,4 +146,99 @@ private static <T> void retryOperation(final CompletableFuture<T> resultFuture,
       resultFuture.whenComplete((t, throwable) -> operationFuture.cancel(false));
     }
   }
+
+  /**
+   * @param <T>
+   * @param resultFuture
+   * @param operation
+   * @param retries
+   * @param retryDelay
+   * @param retryPredicate
+   * @param retryDelayFromException
+   * @param scheduler
+   */
+  private static <T> void retryOperationWithDelay(final CompletableFuture<T> resultFuture,
+      final Supplier<CompletableFuture<T>> operation, final int retries, Duration retryDelay,
+      final Predicate<Throwable> retryPredicate, Function<Throwable, Integer> retryDelayFromException,
+      final ScheduledExecutorService scheduler) {
+
+    if (!resultFuture.isDone()) {
+      final CompletableFuture<T> operationResultFuture = operation.get();
+
+      operationResultFuture.whenComplete((t, throwable) -> {
+        if (throwable != null) {
+          if (throwable instanceof CancellationException) {
+            resultFuture.completeExceptionally(throwable);
+          } else {
+            throwable = extractFutureExceptionCause(throwable);
+            if (!retryPredicate.test(throwable)) {
+              resultFuture.completeExceptionally(throwable);
+            } else if (retries > 0) {
+              long retryDelayFromExceptionMs = retryDelayFromException.apply(throwable);
+              long retryDelayMillis = retryDelayFromExceptionMs > 0 ? retryDelayFromExceptionMs : retryDelay.toMillis();
+              if (scheduler == null) {
+                if (retryDelayMillis > 0) {
+                  logger.warn("No scheduler for delay execution. waitTimeMs: {}", retryDelayMillis);
+                  retryOperationWithDelay(resultFuture, operation, retries - 1, retryDelay, retryPredicate,
+                      retryDelayFromException, scheduler);
+                }
+              } else {
+                final ScheduledFuture<?> scheduledFuture = scheduler.schedule(
+                    () -> retryOperationWithDelay(resultFuture, operation, retries - 1, retryDelay, retryPredicate,
+                        retryDelayFromException, scheduler), retryDelayMillis, MILLISECONDS);
+
+                resultFuture.whenComplete((innerT, innerThrowable) -> scheduledFuture.cancel(false));
+              }
+            } else {
+              resultFuture.completeExceptionally(throwable);
+            }
+          }
+        } else {
+          resultFuture.complete(t);
+        }
+      });
+
+      resultFuture.whenComplete((t, throwable) -> operationResultFuture.cancel(false));
```

Comment:
I don't see a case where resultFuture would be completed when operationResultFuture is not completed.